src/parser.cpp

```diff
@@ -8,6 +8,14 @@
 
 #include <cassert>
 
+#define HANDLE(writer)                                                         \
+	if (handle_error(result, writer))                                      \
```

Comment:
i dont like the name `HANDLE`, maybe something more descriptive, like `HANDLE_ERROR` or `CHECK_AND_RETURN`, like we use elsewhere, would be better.